KenticoOnboardingApplication/src/KenticoOnboardingApplication.Api/Controllers/ListController.cs

```diff
@@ -17,38 +18,153 @@ public class ListController : ApiController
+            ValidateTextAndDateTimes(value);
+            ValidateId(value.Id, shouldBeEmpty: true);
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            var item = await _createItemService.CreateItemAsync(value);
+            var uri = _urlLocator.GetListItemUri(item.Id);
 
             return Created(uri, item);
         }
 
         [Route("{id:guid}")]
-        public async Task<IHttpActionResult> PutItemAsync(Guid id, [FromBody] Item value) =>
-            Ok(await _repository.UpdateItemAsync(value));
+        public async Task<IHttpActionResult> PutItemAsync(Guid id, [FromBody] Item value)
+        {
+            ValidateTextAndDateTimes(value);
+            ValidateId(value.Id, shouldBeEmpty: false);
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            var result = await _updateItemService.UpdateItemAsync(value);
+            if (!result.WasFound)
+            {
+                value.Id = Guid.Empty;
+                return await PostItemAsync(value);
+            }
+
+            return Ok(result.Item);
+        }
 
         [Route("{id:guid}")]
         public async Task<IHttpActionResult> DeleteItemAsync(Guid id)
         {
+            ValidateId(id, shouldBeEmpty: false);
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            var retrievedItem = await _getItemService.GetItemAsync(id);
+            if (!retrievedItem.WasFound)
+            {
+                return NotFound();
+            }
+
             await _repository.DeleteItemAsync(id);
 
             return StatusCode(HttpStatusCode.NoContent);
         }
+
+        private void ValidateId(Guid id, bool shouldBeEmpty)
+        {
+            if (shouldBeEmpty)
+            {
+                ValidateEmptyId(id);
+            }
+            else
+            {
+                ValidateNonEmptyId(id);
+            }
+        }
+
+        private void ValidateEmptyId(Guid id)
+        {
+            if (id != Guid.Empty)
+            {
+                ModelState.AddModelError(nameof(Item.Id), "Id must be empty.");
+            }
+        }
+
+        private void ValidateNonEmptyId(Guid id)
+        {
+            if (id == Guid.Empty)
+            {
+                ModelState.AddModelError(nameof(Item.Id), "Id cannot be empty.");
+            }
+        }
+
+        private void ValidateTextAndDateTimes(Item item)
+        {
+            ValidateText(item.Text);
+            ValidateLastUpdateTime(item.LastUpdateTime);
+            ValidateCreationTime(item.CreationTime);
+        }
+
+        private void ValidateText(string text)
+        {
+            if (string.IsNullOrEmpty(text))
+            {
+                ModelState.AddModelError(nameof(Item.Text), "Text cannot be empty.");
```

Comment:
```suggestion
                ModelState.AddModelError(nameof(Item.Text), $"nameof(Item.Text) cannot be empty.");
```
– I mean, if you rename the property, your message would become outdated. It is maybe a bit more apparent in the two methods below…

Alternatively, you can devise a more life-user-friendly message and dodge the string interpolation by using general English words:
```suggestion
                ModelState.AddModelError(nameof(Item.Text), "Item's text cannot be empty.");
```
(it might be better seen with a less-generic property name:
```
   ModelState.AddModelError(nameof(Item.CreationTime), "The time of item's creation must not be set.");
// -- or even --
   ModelState.AddModelError(nameof(Item.CreationTime), "Don't provide the time of item's origin.");
```
)